### PR TITLE
fix(slack-notifications): [nan-1164] multiple notifications on refresh error

### PR DESCRIPTION
**Pull request by khaliqgant**
## Describe your changes
This annoying issue is hard to reproduce without removing the locking delay on local for the `tryAcquire` method

```
const ttlInMs = 1;
const acquisitionTimeoutMs = 1;
const { tries } = await this.locking.tryAcquire(lockKey, ttlInMs, acquisitionTimeoutMs);
```
With that multiple slack notifications can be seen which in my testing was mitigated by:
- Adding in transactions and `forUpdate` to the database lookups and inserts
- Adding a slight randomized delay for webhook firing to prevent concurrent firing
- `await` the slack notification call

## Testing
- It seemed the customers issue was stemming from proxy calls and only with salesforce which is likely caused by the introspection call. So I tested with a script that made many concurrent proxy calls:

```
Promise.allSettled([
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
  nango.get({
    providerConfigKey: "salesforce",
    connectionId: "sf-2",
    endpoint: `/services/data/v51.0/sobjects/Contacts/describe`,
  }),
```

With this I was seeing multiple slack notifications but with the update logic now only one is received consistently. 

## Issue ticket number and link
NAN-1163

## Checklist before requesting a review (skip if just adding/editing APIs & templates)
- [ ] I added tests, otherwise the reason is: 
- [ ] I added observability, otherwise the reason is:
- [ ] I added analytics, otherwise the reason is:

### CODE REVIEW

File: packages/server/lib/hooks/hooks.ts [hunk 1]
```diff
@@ -185,7 +187,9 @@ export const connectionRefreshFailed = async ({
 
     const slackNotificationService = new SlackService({ orchestratorClient: getOrchestratorClient(), logContextGetter });
 
-    void slackNotificationService.reportFailure(connection, connection.connection_id, 'auth', logCtx.id, environment.id, config.provider);
+    const randomDelay = Math.random() * 1000;
+    await delay(randomDelay);
```
Comment (bodinsamuel) [inline packages/server/lib/hooks/hooks.ts]: seems unnecessary if you wrap the whole hook inside a transaction

Comment (khaliqgant) [reply packages/server/lib/hooks/hooks.ts]: Tested using wrapping the hook inside a transaction and it doesn't fix the multiple slack notification issue

**Comment by bodinsamuel** [comment]
Reproduced locally, you just trying to lock a row that doesn't exists, `forUpdate` will only lock if there is something to lock and there is no unique key constraint so it won't fail to insert two rows. With an advisory lock it seems better

```diff
diff --git a/packages/server/lib/hooks/hooks.integration.test.ts b/packages/server/lib/hooks/hooks.integration.test.ts
new file mode 100644
index 00000000..7404f3a2
--- /dev/null
+++ b/packages/server/lib/hooks/hooks.integration.test.ts
@@ -0,0 +1,139 @@
+import { describe, it, beforeAll } from 'vitest';
+import { multipleMigrations } from '@nangohq/database';
+import { connectionRefreshFailed } from './hooks.js';
+import { configService, connectionService, seeders } from '@nangohq/shared';
+import { logContextGetter } from '@nangohq/logs';
+
+describe('dff', () => {
+    beforeAll(async () => {
+        await multipleMigrations();
+    });
+
+    it('df', async () => {
+        const { env } = await seeders.seedAccountEnvAndUser();
+        await seeders.createConfigSeed(env, 'unauthenticated', 'unauthenticated');
+        const logCtx = await logContextGetter.create({ operation: { type: 'auth', action: 'refresh_token' }, message: 'Token refresh error' }, {} as any, {
+            dryRun: true
+        });
+        const tmp = await seeders.createConnectionSeed(env, 'unauthenticated');
+
+        const { response: connection } = await connectionService.getConnection(tmp.connection_id, 'unauthenticated', env.id);
+
+        const config = await configService.getProviderConfig(connection!.provider_config_key, env.id);
+        const template = configService.getTemplate(config!.provider);
+        await Promise.all([
+            connectionRefreshFailed({
+                authError: { description: 't', type: 'auth' },
+                template,
+                config: config!,
+                logCtx,
+                connection: connection!,
+                environment: env
+            }),
+            connectionRefreshFailed({
+                authError: { description: 't', type: 'auth' },
+                template,
+                config: config!,
+                logCtx,
+                connection: connection!,
+                environment: env
+            }),
+            connectionRefreshFailed({
+                authError: { description: 't', type: 'auth' },
+                template,
+                config: config!,
+                logCtx,
+                connection: connection!,
+                environment: env
+            }),
+            connectionRefreshFailed({
+                authError: { description: 't', type: 'auth' },
+                template,
+                config: config!,
+                logCtx,
+                connection: connection!,
+                environment: env
+            }),
+            connectionRefreshFailed({
+                authError: { description: 't', type: 'auth' },
+                template,
+                config: config!,
+                logCtx,
+                connection: connection!,
+                environment: env
+            }),
+            connectionRefreshFailed({
+                authError: { description: 't', type: 'auth' },
+                template,
+                config: config!,
+                logCtx,
+                connection: connection!,
+                environment: env
+            }),
+            connectionRefreshFailed({
+                authError: { description: 't', type: 'auth' },
+                template,
+                config: config!,
+                logCtx,
+                connection: connection!,
+                environment: env
+            }),
+            connectionRefreshFailed({
+                authError: { description: 't', type: 'auth' },
+                template,
+                config: config!,
+                logCtx,
+                connection: connection!,
+                environment: env
+            }),
+            connectionRefreshFailed({
+                authError: { description: 't', type: 'auth' },
+                template,
+                config: config!,
+                logCtx,
+                connection: connection!,
+                environment: env
+            }),
+            connectionRefreshFailed({
+                authError: { description: 't', type: 'auth' },
+                template,
+                config: config!,
+                logCtx,
+                connection: connection!,
+                environment: env
+            }),
+            connectionRefreshFailed({
+                authError: { description: 't', type: 'auth' },
+                template,
+                config: config!,
+                logCtx,
+                connection: connection!,
+                environment: env
+            }),
+            connectionRefreshFailed({
+                authError: { description: 't', type: 'auth' },
+                template,
+                config: config!,
+                logCtx,
+                connection: connection!,
+                environment: env
+            }),
+            connectionRefreshFailed({
+                authError: { description: 't', type: 'auth' },
+                template,
+                config: config!,
+                logCtx,
+                connection: connection!,
+                environment: env
+            }),
+            connectionRefreshFailed({
+                authError: { description: 't', type: 'auth' },
+                template,
+                config: config!,
+                logCtx,
+                connection: connection!,
+                environment: env
+            })
+        ]);
+    });
+});
diff --git a/packages/server/lib/hooks/hooks.ts b/packages/server/lib/hooks/hooks.ts
index accc1679..3cc4ab89 100644
--- a/packages/server/lib/hooks/hooks.ts
+++ b/packages/server/lib/hooks/hooks.ts
@@ -37,8 +37,6 @@ import { sendAuth as sendAuthWebhook } from '@nangohq/webhooks';
 const logger = getLogger('hooks');
 const orchestrator = getOrchestrator();
 
-const delay = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
-
 export const connectionCreationStartCapCheck = async ({
     providerConfigKey,
     environmentId,
@@ -187,8 +185,6 @@ export const connectionRefreshFailed = async ({
 
     const slackNotificationService = new SlackService({ orchestratorClient: getOrchestratorClient(), logContextGetter });
 
-    const randomDelay = Math.random() * 1000;
-    await delay(randomDelay);
     await slackNotificationService.reportFailure(connection, connection.connection_id, 'auth', logCtx.id, environment.id, config.provider);
 };
 
diff --git a/packages/shared/lib/services/notification/slack.service.ts b/packages/shared/lib/services/notification/slack.service.ts
index e31aa2ce..c2b28eac 100644
--- a/packages/shared/lib/services/notification/slack.service.ts
+++ b/packages/shared/lib/services/notification/slack.service.ts
@@ -9,6 +9,7 @@ import accountService from '../account.service.js';
 import type { LogContext, LogContextGetter } from '@nangohq/logs';
 import type { OrchestratorClientInterface } from '../../clients/orchestrator.js';
 import { Orchestrator } from '../../clients/orchestrator.js';
+import { NangoError } from '../../utils/error.js';
 
 const logger = getLogger('SlackService');
 const TABLE = dbNamespace + 'slack_notifications';
@@ -92,7 +93,11 @@ export class SlackService {
      * notification to the Nango admin account
      */
     private async getNangoAdminConnection(): Promise<NangoConnection | null> {
-        const info = await accountService.getAccountAndEnvironmentIdByUUID(this.nangoAdminUUID as string, this.env);
+        if (!this.nangoAdminUUID) {
+            return null;
+        }
+
+        const info = await accountService.getAccountAndEnvironmentIdByUUID(this.nangoAdminUUID, this.env);
 
         const { success, response: slackConnection } = await connectionService.getConnection(
             this.adminConnectionId,
@@ -107,8 +112,12 @@ export class SlackService {
         return slackConnection;
     }
 
-    private async getAdminEnvironmentId(): Promise<number> {
-        const info = await accountService.getAccountAndEnvironmentIdByUUID(this.nangoAdminUUID as string, this.env);
+    private async getAdminEnvironmentId(): Promise<number | null> {
+        if (!this.nangoAdminUUID) {
+            return null;
+        }
+
+        const info = await accountService.getAccountAndEnvironmentIdByUUID(this.nangoAdminUUID, this.env);
 
         return info?.environmentId as number;
     }
@@ -216,6 +225,10 @@ export class SlackService {
         if (success && !slackNotificationStatus) {
             return;
         }
+        if (!success) {
+            logger.error('failed to create row');
+            return;
+        }
 
         const account = await environmentService.getAccountFromEnvironment(environment_id);
         if (!account) {
@@ -224,6 +237,10 @@ export class SlackService {
 
         const slackConnectionId = generateSlackConnectionId(account.uuid, envName);
         const nangoEnvironmentId = await this.getAdminEnvironmentId();
+        if (!nangoEnvironmentId) {
+            logger.error('No admin environment id');
+            return;
+        }
 
         // we get the connection on the nango admin account to be able to send the notification
         const {
@@ -379,6 +396,11 @@ export class SlackService {
         }
 
         const nangoEnvironmentId = await this.getAdminEnvironmentId();
+        if (!nangoEnvironmentId) {
+            logger.error('No admin environment id');
+            return;
+        }
+
         const slackConnectionId = generateSlackConnectionId(account.uuid, envName);
         const { success: connectionSuccess, response: slackConnection } = await connectionService.getConnection(
             slackConnectionId,
@@ -489,6 +511,13 @@ export class SlackService {
      */
     async addFailingConnection(nangoConnection: NangoConnection, name: string, type: string): Promise<ServiceResponse<NotificationResponse>> {
         return await db.knex.transaction(async (trx) => {
+            // we need to acquire a Lock that prevents any other duplicate hook to create the same row
+            // TODO: create correct unique ID
+            const { rows } = await trx.raw<{ rows: { pg_try_advisory_xact_lock: boolean }[] }>(`SELECT pg_try_advisory_xact_lock(?);`, [34]);
+            if (!rows || rows.length <= 0 || !rows[0]!.pg_try_advisory_xact_lock) {
+                return { success: false, error: new NangoError('adv_lock'), response: null };
+            }
+
             const isOpen = await this.hasOpenNotification(nangoConnection, name, type, trx);
 
             if (isOpen && type === 'auth') {
```

File: packages/shared/lib/services/notification/slack.service.ts [hunk 2]
```diff
@@ -470,6 +472,7 @@ export class SlackService {
                 type,
                 connection_list: [nangoConnection.id as number]
             })
+            .forUpdate()
```
Comment (TBonnin) [inline packages/shared/lib/services/notification/slack.service.ts]: this is a INSERT so `forUpdate` doesn't apply. Not sure why knex allows it

File: packages/shared/lib/services/notification/slack.service.ts [hunk 3]
```diff
@@ -485,35 +488,60 @@ export class SlackService {
...
+                };
+            }
 
-        const { id, connection_list } = isOpen;
+            connection_list.push(nangoConnection.id as number);
+
+            await trx
+                .from<SlackNotification>(TABLE)
+                .where({ id: id as number })
+                .forUpdate()
```
Comment (TBonnin) [inline packages/shared/lib/services/notification/slack.service.ts]: this is an UPDATE so forUpdate doesn't apply

File: packages/shared/lib/services/notification/slack.service.ts [hunk 4]
```diff
@@ -565,52 +571,55 @@ export class SlackService {
...
-            admin_slack_timestamp as string,
-            connection_list.length
-        );
+            logger.info(`Resolving ${nangoConnection.id} type:${type} name:${name}`);
+
+            connection_list.splice(index, 1);
+
+            await trx
+                .from<SlackNotification>(TABLE)
+                .forUpdate()
```
Comment (TBonnin) [inline packages/shared/lib/services/notification/slack.service.ts]: same comment about forUpdate

File: packages/server/lib/hooks/hooks.ts (hunk 1 above)
Comment (TBonnin) [inline packages/server/lib/hooks/hooks.ts]: Adding some delay makes the problem less likely to happen but it doesn't solve the underlying issue

File: packages/shared/lib/services/notification/slack.service.ts [hunk 5]
```diff
@@ -485,35 +488,60 @@ export class SlackService {
      * and if so add the connection id to the connection list.
      */
     async addFailingConnection(nangoConnection: NangoConnection, name: string, type: string): Promise<ServiceResponse<NotificationResponse>> {
-        const isOpen = await this.hasOpenNotification(nangoConnection, name, type);
+        return await db.knex.transaction(async (trx) => {
+            const isOpen = await this.hasOpenNotification(nangoConnection, name, type, trx);
```
Comment (TBonnin) [inline packages/shared/lib/services/notification/slack.service.ts:500]: the problem seems to be a race condition around 
```
const isOpen =  await this.hasOpenNotification(...)
if(!isOpen) {
    await this.createNotification(...)
}
```
where multiple concurrent executions can enter the `if`. A mutex should solve the problem here. What do you think?

**Review by TBonnin — COMMENTED** [review]
adding some delay makes the problem less likely to happen but

**Comment by khaliqgant** [comment]
Thanks both. I’ll update with a lock which agreed should be a more resilient fix

File: packages/shared/lib/services/notification/slack.service.ts (hunk 2 above)
Comment (khaliqgant) [reply packages/shared/lib/services/notification/slack.service.ts]: Updated

File: packages/shared/lib/services/notification/slack.service.ts (hunk 4 above)
Comment (khaliqgant) [reply packages/shared/lib/services/notification/slack.service.ts]: Updated

File: packages/server/lib/hooks/hooks.ts (hunk 1 above)
Comment (khaliqgant) [reply packages/server/lib/hooks/hooks.ts]: True, updated to use mutex

File: packages/shared/lib/services/notification/slack.service.ts (hunk 5 above)
Comment (khaliqgant) [reply packages/shared/lib/services/notification/slack.service.ts:500]: Yup, updated.

**Review by TBonnin — APPROVED** [review]

**Review by bodinsamuel — APPROVED** [review]

**Pull request merged**